File: frontend/src/api/realms.ts
import { apiGet, apiDelete, apiUpload } from './client';
import type { Realm, Homework, FileContent, FileEntry } from '@/types';

export function listRealms(): Promise<Realm[]> {
  return apiGet<Realm[]>('/realms');
}

export function getRealm(id: string): Promise<Realm> {
  return apiGet<Realm>(`/realms/${id}`);
}

export function getHomework(
  realmId: string,
  hwId: string,
): Promise<Homework & { student_files: FileEntry[]; gold_files: FileEntry[] }> {
  return apiGet(`/realms/${realmId}/homeworks/${hwId}`);
}

export function uploadRealm(file: File, name: string): Promise<Realm> {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('name', name);
  return apiUpload<Realm>('/realms/upload', formData);
}

export function deleteRealm(id: string): Promise<void> {
  return apiDelete(`/realms/${id}`);
}

export function getHomeworkFile(realmId: string, hwId: string, filePath: string): Promise<FileContent> {
  return apiGet<FileContent>(`/realms/${realmId}/homeworks/${hwId}/files/${filePath}`);
}

export function uploadGoldFile(
  realmId: string,
  hwId: string,
  file: File,
): Promise<{ homework_id: string; path: string; filename: string }> {
  const formData = new FormData();
  formData.append('file', file);
  return apiUpload(`/realms/${realmId}/homeworks/${hwId}/gold`, formData);
}

File: backend/app/api/realms.py
from fastapi import APIRouter, File, Form, UploadFile

from app.schemas import HomeworkRead, RealmRead
from app.services import storage_manager_client as sm

router = APIRouter(prefix="/realms", tags=["realms"])


@router.post("/upload", response_model=RealmRead, status_code=201)
async def upload_realm(
    file: UploadFile = File(...),
    name: str = Form(...),
):
    return await sm.upload_realm(file, name)


@router.get("", response_model=list[RealmRead])
def list_realms():
    return sm.list_realms()


@router.get("/{realm_id}", response_model=RealmRead)
def get_realm(realm_id: str):
    return sm.get_realm(realm_id)


@router.get("/{realm_id}/homeworks/{homework_id}", response_model=HomeworkRead)
def get_homework(realm_id: str, homework_id: str):
    return sm.get_homework_detail(realm_id, homework_id)


@router.get("/{realm_id}/homeworks/{homework_id}/files/{file_path:path}")
def get_homework_file(realm_id: str, homework_id: str, file_path: str):
    return sm.get_file_content(realm_id, homework_id, file_path)


@router.post("/{realm_id}/homeworks/{homework_id}/gold", status_code=201)
async def upload_gold_file(realm_id: str, homework_id: str, file: UploadFile = File(...)):
    return await sm.upload_gold_file(realm_id, homework_id, file)


@router.delete("/{realm_id}", status_code=204)
def delete_realm(realm_id: str):
    sm.delete_realm(realm_id)
